File: frontend/src/lib/searchSort.ts
import type { CryptCard, CryptSort } from './cryptSearch'
import type { LibraryCard, LibrarySort } from './librarySearch'

const byName = (left: string, right: string) => left.localeCompare(right, 'en')

export function sortCryptResults<T extends CryptCard>(cards: T[], sort: CryptSort): T[] {
  return [...cards].sort((left, right) => {
    switch (sort) {
      case 'capacity_asc':
        return left.capacity - right.capacity || byName(left.name, right.name)
      case 'clan':
        return (
          byName(left.clan, right.clan) ||
          right.capacity - left.capacity ||
          byName(left.name, right.name)
        )
      case 'group':
        return left.grp - right.grp || right.capacity - left.capacity || byName(left.name, right.name)
      case 'name':
        return byName(left.name, right.name)
      case 'sect':
        return (
          byName(left.sect ?? '', right.sect ?? '') ||
          right.capacity - left.capacity ||
          byName(left.name, right.name)
        )
      case 'capacity_desc':
        return right.capacity - left.capacity || byName(left.name, right.name)
    }
  })
}

function numericCost(value: string | null): number | null {
  if (!value || !/^\d+$/.test(value)) return null
  return Number(value)
}

function compareCost(left: string | null, right: string | null, direction: 1 | -1): number {
  const leftNumber = numericCost(left)
  const rightNumber = numericCost(right)
  if (leftNumber !== null && rightNumber === null) return -1
  if (leftNumber === null && rightNumber !== null) return 1
  if (leftNumber === null || rightNumber === null) return 0
  return (leftNumber - rightNumber) * direction
}

function compareLibraryType(left: LibraryCard, right: LibraryCard): number {
  const leftDisciplines = [...left.disciplines].sort().join(',')
  const rightDisciplines = [...right.disciplines].sort().join(',')
  return (
    byName(left.types.join('/'), right.types.join('/')) ||
    byName(left.clan ?? '', right.clan ?? '') ||
    byName(leftDisciplines, rightDisciplines) ||
    byName(left.name, right.name)
  )
}

function compareLibraryRequirement(left: LibraryCard, right: LibraryCard): number {
  const leftHasClan = left.clan ? 0 : 1
  const rightHasClan = right.clan ? 0 : 1
  const leftHasDiscipline = left.disciplines.length > 0 ? 0 : 1
  const rightHasDiscipline = right.disciplines.length > 0 ? 0 : 1
  const leftDisciplines = [...left.disciplines].sort().join(',')
  const rightDisciplines = [...right.disciplines].sort().join(',')
  return (
    leftHasClan - rightHasClan ||
    byName(left.clan ?? '', right.clan ?? '') ||
    leftHasDiscipline - rightHasDiscipline ||
    byName(leftDisciplines, rightDisciplines) ||
    compareLibraryType(left, right)
  )
}

export function sortLibraryResults<T extends LibraryCard>(cards: T[], sort: LibrarySort): T[] {
  return [...cards].sort((left, right) => {
    switch (sort) {
      case 'requirement':
        return compareLibraryRequirement(left, right)
      case 'cost_desc':
        return (
          compareCost(left.blood_cost, right.blood_cost, -1) ||
          compareCost(left.pool_cost, right.pool_cost, -1) ||
          compareLibraryType(left, right)
        )
      case 'cost_asc':
        return (
          compareCost(left.blood_cost, right.blood_cost, 1) ||
          compareCost(left.pool_cost, right.pool_cost, 1) ||
          compareLibraryType(left, right)
        )
      case 'type':
        return compareLibraryType(left, right)
      case 'name':
        return byName(left.name, right.name)
    }
  })
}

File: core/src/search_sort.rs
//! Stable VDB-compatible card ordering shared by exact and semantic search.

use std::cmp::Ordering;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CryptSort {
    CapacityDesc,
    CapacityAsc,
    Clan,
    Group,
    Name,
    Sect,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CryptSortRecord {
    pub id: u32,
    pub name_ascii: String,
    pub clan: String,
    pub capacity: i64,
    pub group: i64,
    pub sect: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LibrarySort {
    Requirement,
    CostDesc,
    CostAsc,
    Name,
    Type,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LibrarySortRecord {
    pub id: u32,
    pub name_ascii: String,
    pub types: Vec<String>,
    pub clan: String,
    pub disciplines: Vec<String>,
    pub blood_cost: Option<String>,
    pub pool_cost: Option<String>,
}

fn compare_text(left: &str, right: &str) -> Ordering {
    left.bytes()
        .map(|byte| byte.to_ascii_lowercase())
        .cmp(right.bytes().map(|byte| byte.to_ascii_lowercase()))
}

fn compare_name(left_id: u32, left: &str, right_id: u32, right: &str) -> Ordering {
    compare_text(left, right).then_with(|| left_id.cmp(&right_id))
}

pub fn compare_crypt(left: &CryptSortRecord, right: &CryptSortRecord, sort: CryptSort) -> Ordering {
    let name = || compare_name(left.id, &left.name_ascii, right.id, &right.name_ascii);
    match sort {
        CryptSort::CapacityDesc => right.capacity.cmp(&left.capacity).then_with(name),
        CryptSort::CapacityAsc => left.capacity.cmp(&right.capacity).then_with(name),
        CryptSort::Clan => compare_text(&left.clan, &right.clan)
            .then_with(|| right.capacity.cmp(&left.capacity))
            .then_with(name),
        CryptSort::Group => left
            .group
            .cmp(&right.group)
            .then_with(|| right.capacity.cmp(&left.capacity))
            .then_with(name),
        CryptSort::Name => name(),
        CryptSort::Sect => compare_text(&left.sect, &right.sect)
            .then_with(|| right.capacity.cmp(&left.capacity))
            .then_with(name),
    }
}

fn sorted_join(values: &[String], separator: &str) -> String {
    let mut values = values.to_vec();
    values.sort_by(|left, right| compare_text(left, right));
    values.join(separator)
}

fn compare_optional_requirement(left: &str, right: &str) -> Ordering {
    left.is_empty()
        .cmp(&right.is_empty())
        .then_with(|| compare_text(left, right))
}

fn numeric_cost(value: &Option<String>) -> Option<u64> {
    value
        .as_deref()
        .filter(|value| !value.is_empty() && value.bytes().all(|byte| byte.is_ascii_digit()))
        .and_then(|value| value.parse().ok())
}

fn compare_cost(left: &Option<String>, right: &Option<String>, descending: bool) -> Ordering {
    match (numeric_cost(left), numeric_cost(right)) {
        (Some(left), Some(right)) if descending => right.cmp(&left),
        (Some(left), Some(right)) => left.cmp(&right),
        (Some(_), None) => Ordering::Less,
        (None, Some(_)) => Ordering::Greater,
        (None, None) => Ordering::Equal,
    }
}

fn compare_library_type(left: &LibrarySortRecord, right: &LibrarySortRecord) -> Ordering {
    let left_types = left.types.join("/");
    let right_types = right.types.join("/");
    let left_disciplines = sorted_join(&left.disciplines, ",");
    let right_disciplines = sorted_join(&right.disciplines, ",");
    compare_text(&left_types, &right_types)
        .then_with(|| compare_optional_requirement(&left.clan, &right.clan))
        .then_with(|| compare_optional_requirement(&left_disciplines, &right_disciplines))
        .then_with(|| compare_name(left.id, &left.name_ascii, right.id, &right.name_ascii))
}

fn compare_library_requirement(left: &LibrarySortRecord, right: &LibrarySortRecord) -> Ordering {
    let left_disciplines = sorted_join(&left.disciplines, ",");
    let right_disciplines = sorted_join(&right.disciplines, ",");
    compare_optional_requirement(&left.clan, &right.clan)
        .then_with(|| compare_optional_requirement(&left_disciplines, &right_disciplines))
        .then_with(|| compare_library_type(left, right))
}

pub fn compare_library(
    left: &LibrarySortRecord,
    right: &LibrarySortRecord,
    sort: LibrarySort,
) -> Ordering {
    match sort {
        LibrarySort::Requirement => compare_library_requirement(left, right),
        LibrarySort::CostDesc => compare_cost(&left.blood_cost, &right.blood_cost, true)
            .then_with(|| compare_cost(&left.pool_cost, &right.pool_cost, true))
            .then_with(|| compare_library_type(left, right)),
        LibrarySort::CostAsc => compare_cost(&left.blood_cost, &right.blood_cost, false)
            .then_with(|| compare_cost(&left.pool_cost, &right.pool_cost, false))
            .then_with(|| compare_library_type(left, right)),
        LibrarySort::Name => compare_name(left.id, &left.name_ascii, right.id, &right.name_ascii),
        LibrarySort::Type => compare_library_type(left, right),
    }
}

pub fn crypt_order(records: &[CryptSortRecord], sort: CryptSort) -> Vec<u32> {
    let mut records = records.iter().collect::<Vec<_>>();
    records.sort_by(|left, right| compare_crypt(left, right, sort));
    records.into_iter().map(|record| record.id).collect()
}

pub fn library_order(records: &[LibrarySortRecord], sort: LibrarySort) -> Vec<u32> {
    let mut records = records.iter().collect::<Vec<_>>();
    records.sort_by(|left, right| compare_library(left, right, sort));
    records.into_iter().map(|record| record.id).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn crypt(
        id: u32,
        name: &str,
        clan: &str,
        capacity: i64,
        group: i64,
        sect: &str,
    ) -> CryptSortRecord {
        CryptSortRecord {
            id,
            name_ascii: name.to_owned(),
            clan: clan.to_owned(),
            capacity,
            group,
            sect: sect.to_owned(),
        }
    }

    fn library(
        id: u32,
        name: &str,
        card_type: &str,
        clan: &str,
        disciplines: &[&str],
        blood: Option<&str>,
        pool: Option<&str>,
    ) -> LibrarySortRecord {
        LibrarySortRecord {
            id,
            name_ascii: name.to_owned(),
            types: vec![card_type.to_owned()],
            clan: clan.to_owned(),
            disciplines: disciplines
                .iter()
                .map(|value| (*value).to_owned())
                .collect(),
            blood_cost: blood.map(str::to_owned),
            pool_cost: pool.map(str::to_owned),
        }
    }

    #[test]
    fn crypt_modes_have_stable_tie_breaks() {
        let cards = vec![
            crypt(3, "Zulu", "Ventrue", 5, 7, "Camarilla"),
            crypt(2, "alpha", "Banu Haqim", 5, 6, "Camarilla"),
            crypt(1, "Alpha", "Ventrue", 6, 6, "Anarch"),
        ];
        assert_eq!(crypt_order(&cards, CryptSort::CapacityDesc), vec![1, 2, 3]);
        assert_eq!(crypt_order(&cards, CryptSort::Clan), vec![2, 1, 3]);
        assert_eq!(crypt_order(&cards, CryptSort::Group), vec![1, 2, 3]);
        assert_eq!(crypt_order(&cards, CryptSort::Sect), vec![1, 2, 3]);
    }

    #[test]
    fn library_requirements_sort_present_values_before_empty() {
        let cards = vec![
            library(1, "None", "Action", "", &[], None, None),
            library(2, "Clan", "Action", "Ventrue", &[], None, None),
            library(3, "Discipline", "Action", "", &["dom"], None, None),
        ];
        assert_eq!(
            library_order(&cards, LibrarySort::Requirement),
            vec![2, 3, 1]
        );
        assert_eq!(library_order(&cards, LibrarySort::Type), vec![2, 3, 1]);
    }

    #[test]
    fn numeric_costs_precede_variable_and_absent_costs() {
        let cards = vec![
            library(1, "X", "Combat", "", &[], Some("X"), None),
            library(2, "Two", "Combat", "", &[], Some("2"), None),
            library(3, "One", "Combat", "", &[], Some("1"), None),
            library(4, "None", "Combat", "", &[], None, None),
        ];
        assert_eq!(
            library_order(&cards, LibrarySort::CostAsc),
            vec![3, 2, 4, 1]
        );
        assert_eq!(
            library_order(&cards, LibrarySort::CostDesc),
            vec![2, 3, 4, 1]
        );
    }
}
